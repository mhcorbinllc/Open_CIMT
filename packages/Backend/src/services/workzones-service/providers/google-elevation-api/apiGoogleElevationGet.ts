import {guid} from "dyna-guid";
import {dynaError} from "dyna-error";

import {apiFetch} from "../../../../utils/apiFetch";

export interface IGeoPoint {
  lat: number;
  lng: number;
  alt?: number;
}

interface IGoogleElevationsResponse {
  results: {
    elevation: number;
    location: {
      lat: number;
      lng: number;
    };
    resolution: number;
  }[];
  status: 'ok' | any;
  error_message?: string;
}

export const apiGoogleElevationGet = async (positions: IGeoPoint[]): Promise<IGeoPoint[]> => {
  let response: IGoogleElevationsResponse | null = null;

  try {
    response = await apiFetch<{locations: string; key: string}, void, IGoogleElevationsResponse>(
      {
        method: 'GET',
        url: 'https://maps.googleapis.com/maps/api/elevation/json',
        query: {
          locations: positions
            .map((
              {
                lat,
                lng,
              },
            ) => [lat, lng]
              .map(v => v.toString()))
            .join("|"),
          key: process.env.GOOGLE_MAP_ELEVATION_API_KEY || '',
        },
      },
    );
  }
  catch (e) {
    const errorId = guid();
    const error = dynaError({
      code: 202110261119,
      status: 503,
      message: `Google Elevation API service responded error. Check server logs with error id ${errorId}.`,
      parentError: e,
    });
    console.error(error.message, error);
    throw error;
  }
  if (!response) return []; // 4TS

  const {
    results,
    status,
    error_message,
  } = response;

  if (status !== 'OK') {
    const error = dynaError({
      code: 202202150942,
      status: 503,
      message: `Google API Elevation responded with error status [${status}]: ${error_message || 'Unknown error'}. Check server logs with error id ${guid()}.`,
      data: {
        requestedPosition: positions,
        response,
      },
    });
    console.error(error.message, error);
    throw error;
  }

  return results.map(result=>({
    lat: result.location.lat,
    lng: result.location.lng,
    alt: result.elevation,
  }));
};
