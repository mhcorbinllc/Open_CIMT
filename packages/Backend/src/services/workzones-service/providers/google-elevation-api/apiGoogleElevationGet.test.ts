import {apiGoogleElevationGet} from "./apiGoogleElevationGet";

describe('apiGoogleElevationGet', ()=>{
  test('Get elevation for two points', async ()=> {
    const elevations = await apiGoogleElevationGet([
      {
        lng: 48.17835117870872,
        lat: 16.309423872994653,
      },
      {
        lng: 48.19842905647751,
        lat: 16.31575585914963,
      },
    ]);
    expect(elevations).toMatchSnapshot();
  });
});
