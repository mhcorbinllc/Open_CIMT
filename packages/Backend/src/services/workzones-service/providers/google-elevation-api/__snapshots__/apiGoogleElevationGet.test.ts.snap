// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiGoogleElevationGet Get elevation for two points 1`] = `
Array [
  Object {
    "alt": 1123.64013671875,
    "lat": 16.30942387299465,
    "lng": 48.17835117870872,
  },
  Object {
    "alt": 1111.77685546875,
    "lat": 16.31575585914963,
    "lng": 48.19842905647751,
  },
]
`;
